guibot/desktopcontrol.py

```diff
@@ -758,10 +758,9 @@ def mouse_click(self, button=None, count=3, modifiers=None):
         if modifiers != None:
             self.keys_toggle(modifiers, True)
         for _ in range(count):
-            self._backend_obj.mousePress(button)
-            # BUG: the mouse button is pressed down forever (on LEFT)
+            self._backend_obj.mouseDown(button)
```

Comment:
Has this been fixed?? This would be great as the issue was reported some years ago and we couldn't conclude anything about the reason, just the remedy here.

Reply:
Ah but I see the remedy here is to just use mouse down/up. The idea for using the full `mousePress` in the past was that the moment it is fixed we could just drop the BUG and extra sleep + mouse up lines. So we keep both the way the API should be used and additional calls to counter the fact that the proper way didn't work.

Reply:
Oh ok, I went through the commits searching for some answers, I guess it was a good hunch. Perhaps we can keep the `mousePress` commented in the code with
```python
# NOTE:  the default method doesn't support the toggle timeout setting
# self._backend_obj.mousePress(button)
```

Reply:
> The idea for using the full mousePress in the past was that the moment it is fixed we could just drop the BUG and extra sleep + mouse up lines.

Actually I replaced the `mousePress` by a `mouseDown` for consistency with other backends, and also because it might be useful for users to customize the interval between mouse press and release. Some applications might not react well with this and I didn't think it was worth removing it.

I believe you're referring to this issue:
https://github.com/sibson/vncdotool/issues/51

If so, then AFAIK it hasn't been fixed, no. But keeping the commented out lines is a good option.

Reply:
> Actually I replaced the `mousePress` by a `mouseDown` for consistency with other backends, and also because it might be useful for users to customize the interval between mouse press and release. Some applications might not react well with this and I didn't think it was worth removing it.

Yes, there is a consistent "BUG" in all of them about why this is the case :joy: It is so consistent in fact that we may just drop the attempts to use their "click" API calls and simply use mouse down/up throughout, dropping all BUG tags too.

> I believe you're referring to this issue:
> [sibson/vncdotool#51](https://github.com/sibson/vncdotool/issues/51)
> 
> If so, then AFAIK it hasn't been fixed, no. But keeping the commented out lines is a good option.

Alright, up to you. We could either (1) keep the commented calls to the perfect matching methods (waiting for them to at least get a `timeout` argument in future versions) or we could just drop all BUG tags and use mouse down/up throughout.

Reply:
> Yes, there is a consistent "BUG" in all of them about why this is the case joy It is so consistent in fact that we may just drop the attempts to use their "click" API calls and simply use mouse down/up throughout, dropping all BUG tags too.

Indeed, but I don't think it's a problem with vncdotool, but rather with QEMU (even when using vncdotool as the backend). Today I did a few tests to debug this issue, let me share my findings.

I created an HTML file with a simple "CLICK ME" button that logged both the mouse down and the mouse up events and opened it on Firefox in a Windows 10 VM. Then I tested clicking on it 20 times in a row without any `sleep()` calls using vncdotool, qemu monitor (from avocado) and the native win32 API through Python's `ctypes`. This third method was to rule out any possibility of Windows or Firefox itself swallowing events as the Python script was running in the VM locally (nothing remotely as the other two methods). It turns out that the first two methods resulted in missed click events while the third worked flawlessly. I thought of two possibilities: either it's a problem with both vncdotool and QEMU/QEMU monitor, or it's a problem with QEMU only, since the VNC server we connect to *is* actually QEMU itself (QEMU's rfb implementation).

To check which was to blame, first I ran this script with mspaint opened (similarly to what you did in that issue):

```python
from vncdotool import api

backend = api.connect("192.168.255.5:0")

backend.refreshScreen()
for i in range(5, 40):
    for j in range(20, 60):
        backend.mouseMove(10*i, 10*j)
        backend.mousePress(1)
```

This is simple: it uses vncdotool to connect to the VNC port exposed by QEMU on this Windows machine. This resulted in the following image:
![image](https://user-images.githubusercontent.com/13439135/94204522-c7926b80-fe97-11ea-9e94-0b13419bcb98.png)

Then I installed TightVNC server on the Windows VM and replaced line:
`backend = api.connect("192.168.255.5:0")`
with
`backend = api.connect("192.168.5.1::5900", password=None)`

And re-ran. 192.168.5.1 here is the IP of the Windows VM, not of the container. So now I was not connecting to the QEMU VNC port, but rather to the TightVNC server running _inside_ the VM. This was the result I got:

![image](https://user-images.githubusercontent.com/13439135/94204654-01637200-fe98-11ea-98a2-c984841642a5.png)

So it worked perfectly. This leads me to believe that QEMU is to blame. It probably doesn't handle click events sent too fast, be it through their monitor or through their built-in rfb implementation. That's why both of these backends in guibot have this bug: the `QemuDesktopControl` uses QEMU's monitor and `VNCDoToolDesktopControl` is always tested with the QEMU exposed vnc port, never with a server running in the VM.

Reply:
This is a very in-depth study of the situation but in comparison to the ones done before this one is additionally isolating the problem to QEMU. Do you think we should report this to the QEMU project itself? It might be worth to use the collected debugging information to at least report the findings about the bug there.

Reply:
I think so, yes. But I believe more information is required before reporting to the QEMU project; for instance, compiling and testing with the latest stable, testing different displays (e.g. `gtk` vs `sdl`), testing different input devices (e.g. `tablet` vs `ps2`) and so on.

Reply:
:exploding_head: